### ISSUE TITLE
Maximum call stack size exceeded in ReactDOMComponent._updateDOMChildren

### ISSUE THREAD

**Issue by andersekdahl** [facebook/react#8030]
First off, I realize that this likely isn't a problem with React but a problem with our application, but I tought I'd open an issue to see if you think it might be a problem in React, or if you could give any pointers on how to dig deeper. Since the call stack only contains React code, there isn't all that much for us to trouble shoot to see if we've messed up.

**Do you want to request a _feature_ or report a _bug_?**
Possible bug.

**What is the current behavior?**
We sometimes get random "Maximum call stack size exceeded" errors from a small percentage of our users. It doesn't look like it's related to browser/device as it seems to happen on all types of devices from mobile to tablet to desktop and in Firefox, Chrome, Safari (both iOS and Mac).

The stack trace doesn't contain any of our code, only React. It looks like this:

``` js
// at lastProps line 795, column 0 (webpack:///./~/react/lib/ReactDOMComponent.js:795)

      _updateDOMProperties: function (lastProps, nextProps, transaction) { // <-- this line
        var propKey;
        var styleName;
        var styleUpdates;

// at _updateDOMProperties line 758, column 0 // (webpack:///./~/react/lib/ReactDOMComponent.js:758)


        assertValidProps(this, nextProps);
        this._updateDOMProperties(lastProps, nextProps, transaction) // <-- this line
        this._updateDOMChildren(lastProps, nextProps, transaction, context);

        switch (this._tag) {

// at updateComponent line 717, column 0 (webpack:///./~/react/lib/ReactDOMComponent.js:717)

      receiveComponent: function (nextElement, transaction, context) {
        var prevElement = this._currentElement;
        this._currentElement = nextElement;
        this.updateComponent(transaction, prevElement, nextElement, context); // <-- this line
      },


// at receiveComponent line 126, column 0 (webpack:///./~/react/lib/ReactReconciler.js:126)

          ReactRef.detachRefs(internalInstance, prevElement);
        }

        internalInstance.receiveComponent(nextElement, transaction, context); // <-- this line

        if (refsChanged && internalInstance._currentElement && internalInstance._currentElement.ref != null) {
          transaction.getReactMountReady().enqueue(attachRefs, internalInstance);

// at receiveComponent line 110, column 0 (webpack:///./~/react/lib/ReactChildReconciler.js:110)

        var prevElement = prevChild && prevChild._currentElement;
        var nextElement = nextChildren[name];
        if (prevChild != null && shouldUpdateReactComponent(prevElement, nextElement)) {
          ReactReconciler.receiveComponent(prevChild, nextElement, transaction, context); // <-- this line
          nextChildren[name] = prevChild;
        } else {
          if (prevChild) {

// at updateChildren line 215, column 0 (webpack:///./~/react/lib/ReactMultiChild.js:215)

            }
          }
          nextChildren = flattenChildren(nextNestedChildrenElements, selfDebugID);
          ReactChildReconciler.updateChildren(prevChildren, nextChildren, mountImages, removedNodes, transaction, this, this._hostContainerInfo, context, selfDebugID); // <-- this line
          return nextChildren;
        },


// at _reconcilerUpdateChildren line 314, column 0 (webpack:///./~/react/lib/ReactMultiChild.js:314)

        var prevChildren = this._renderedChildren;
        var removedNodes = {};
        var mountImages = [];
        var nextChildren = this._reconcilerUpdateChildren(prevChildren, nextNestedChildrenElements, mountImages, removedNodes, transaction, context); // <-- this line
        if (!nextChildren && !prevChildren) {
          return;
        }

// at _updateChildren line 301, column 0 (webpack:///./~/react/lib/ReactMultiChild.js:301)

        updateChildren: function (nextNestedChildrenElements, transaction, context) {
          // Hook used by React ART
          this._updateChildren(nextNestedChildrenElements, transaction, context); // <-- this line
        },


// at updateChildren line 941, column 0 (webpack:///./~/react/lib/ReactDOMComponent.js:941)

            setContentChildForInstrumentation.call(this, null);
          }

          this.updateChildren(nextChildren, transaction, context); // <-- this line
        }
      },


// at _updateDOMChildren line 759, column 0 (webpack:///./~/react/lib/ReactDOMComponent.js:759)


        assertValidProps(this, nextProps);
        this._updateDOMProperties(lastProps, nextProps, transaction);
        this._updateDOMChildren(lastProps, nextProps, transaction, context); // <-- this line

        switch (this._tag) {
          case 'input':
```

We have some pages that are larger than others, but not anyone that have nesting deep enough that it could realistically cause a maximum call stack size exceeded error.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
15.3.1. Don't know if this is true for other versions, we haven't had any other version deployed.

Any help reagarding this would be much appreciated!

**Comment by syranide** [comment]
IIRC this is often an indication that you're setting state in `componentDIdMount` or some other life-cycle method and for some reason you end up endlessly changing the state (remember that even if you don't actually change any values in state, it will still queue an update unless you have a proper `shouldComponentUpdate`)... until it runs out of stack.

**Comment by andersekdahl** [comment]
Yeah I've looked through most of our code for that but couldn't find any such cases. And if that was the issue, I'd expect our code to be shown in the stack trace and for it to happen a lot more. We have a couple of different stack traces in our logs and our code is never in it.

But it's still the most likely explaination that the error is on our side, so I'll dig more.

**Comment by jrmartin** [comment]
Hello andersekdahl, did you end up finding what was causing your bug? We have encountered the same glitch in our application, and setting state in componentDidMount() doesn't seem to be the issue either.

**Comment by aweary** [comment]
@jrmartin (or @andersekdahl) could you provide a small example reproducing the issue? You can use [this JSFiddle](https://jsfiddle.net/reactjs/69z2wepo/) as a starting point. As mentioned before, it is likely an issue with your application not React, so it's crucial we have evidence showing otherwise.

**Comment by andersekdahl** [comment]
We are still seeing the issue in our logs, but we have never been able to reproduce it. It may be an issue in our code but I can't be sure.

It seems like the error occurs randomly, we haven't seen any direct pattern.

Since I can't reproduce it I know that it's not actionable for you guys. So I'm sorta just hoping that Fiber might magically fix it, or make it obvious what's causing the error.

**Comment by jrmartin** [comment]
Unfortunately there's way too much code in the application to isolate the issue in a codepen/jsfiddle. 
The bug started happening after updating to latest Chrome, with previous Chrome version it was working fine, it's still working fine in Firefox without any issues. We are using Griddle.js in a React Component, and the issue happens when we scroll down a very long list inside the Griddle component. 

This is the screenshot of the stack trace error we are getting:
![stack trace error](https://cloud.githubusercontent.com/assets/4562825/22433016/b83a8de4-e6cb-11e6-988e-dc8ea31b4176.png)

My first impression was that it's a bug in our application, but stack trace doesn't point anywhere in there and I have double checked all states updates in componentDidMount() without luck. Also, that it's working fine in Firefox and previous Chrome version, but not latest, it's making me think it might not be a bug in our application but not sure.

**Comment by DeerMichel** [comment]
+1

**Comment by gaearon** [comment]
There is no need to “+1” this. It never helped anybody.

So far it looks like this is an application bug (rendering components recursively), and there is no evidence to suggest otherwise. If you experience the same problem and want to see it fixed, the best you can do is to publish a project reproducing the problem, and link to this project in this issue.

Since there has been no such reproducing example for a long time, I’m closing this as application bug. Happy to reopen if you can show this is a bug in React.

**Comment by xiankai** [comment]
@andersekdahl @jrmartin I chanced upon this issue while trying to debug a similar error, and I wanted to share how I figured it out.

I started by hiding most of my application (with `{ false && <Component /> }`), until the error disappeared, then slowly toggled them back on one by one. If I found one component that made the error appear again, I would dive deeper into it and repeat the process, until I found the source of the error.

In my case it was code in `componentWillUpdate` that was supposed to normalize some props but was called again and again because the props weren't normalized correctly...

Hope that helps.
